watcher.go

```diff
@@ -0,0 +1,141 @@
+	// attempts reaches Watcher.MaxAttempts
+	ErrMaxAttempts = errors.New("watcher: max attempts reached. giving up")
+)
+
+func (w *Watcher) client() *Client {
+	if client := w.Client; client != nil {
+		return client
+	}
+	// This is copied from DefaultTransport.
+	// we don't do an actual copy because the Trasnport uses mutexes.
+	// copied mutexes are ineffective.
+	// You will panic on concurrent map access if running
+	// with other clients without instantiating a new transport here.
+	ts := &http.Transport{
+		DialContext: (&net.Dialer{
+			Timeout: 5 * time.Second,
+		}).DialContext,
+		DisableCompression:    true,
+		MaxIdleConns:          5,
+		MaxIdleConnsPerHost:   2,
+		IdleConnTimeout:       90 * time.Second,
+		TLSHandshakeTimeout:   5 * time.Second,
+		ResponseHeaderTimeout: 10 * time.Minute,
+		ExpectContinueTimeout: 5 * time.Second,
+	}
+	c := &Client{
+		Transport: ts,
+	}
+	return c
+}
+
+// Watch is the package-level Watch defintion which is called on
+// DefaultWatcher.
+func Watch(ctx context.Context, key string, handler WatcherFunc) {
+	DefaultWatcher.Watch(ctx, key, handler)
+}
+
+// WatchPrefix is the package-level WatchPrefix defintion which is called on
+// DefaultWatcher.
+func WatchPrefix(ctx context.Context, prefix string, handler WatcherFunc) {
+	DefaultWatcher.WatchPrefix(ctx, prefix, handler)
+}
+
+// Watch executes a long poll for changes to the given key.  handler will be
+// called immediately upon registration to initialize the watch and returns
+// the initial value (as a list, this is what Consul API returns).  There is
+// a chance that Watch can miss updates due to the way the API is implemented.
+// See https://github.com/hashicorp/consul/issues/1761 for details.
+func (w *Watcher) Watch(ctx context.Context, key string, handler WatcherFunc) {
+	w.watching(ctx, key, handler, nil)
+}
+
+// WatchPrefix executes a long poll for changes to anything under the given
+// prefix.  handler will be called immediately upon registration to initialize
+// the watch and returns the initial value (as a list, this is what Consul API
+// returns).  There is a chance that WatchPrefix can miss updates due to the
+// way the API is implemented.  See https://github.com/hashicorp/consul/issues/1761
+// for details.
+func (w *Watcher) WatchPrefix(ctx context.Context, prefix string, handler WatcherFunc) {
+	q := Query{Param{
+		Name:  "recurse",
+		Value: "",
+	}}
+	w.watching(ctx, prefix, handler, q)
+}
+
+func (w *Watcher) watching(ctx context.Context, key string, handler WatcherFunc, q Query) {
+	if w.MaxAttempts <= 0 {
+		w.MaxAttempts = 10
+	}
+
+	path := "/v1/kv/" + key
+	value := "0"
+
+	attempt := 0
+	for {
+		q = updateParam(q, Param{Name: "index", Value: value})
+		resp := []KeyData{}
+		hdr, err := w.client().do(ctx, "GET", path, q, nil, &resp)
+		if err != nil {
+			attempt += 1
+			if attempt >= w.MaxAttempts {
+				handler(nil, ErrMaxAttempts)
+				return
```

Comment:
Are you sure this is the behavior we want? I feel like we should just pass the original error to the handler and let it decide what to do, potentially it would cancel the context, then use that as a way to exit the loop:
```go
for ctx.Err() == nil {
    ...
}
```

Reply:
Yeah, we can pass the original error.

I was thinking that we might pass this because the other errors would have went through, but that's only the case if they're not temporary errors.

Reply:
Yes we should keep the timeout code, what I'm less comfortable with is the max-attempts, which may trigger in normal cases where there were no changes, but for some reason the library chooses that it's waited enough and gives up. I feel like it should be the program's responsibility to choose when to give up or whether the error is expected.

Note that retrying on a timeout is not going to happen consul since by definition, we've been waiting for a while when it triggers.